apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/Trace/SimpleManagedTracer.cs

```diff
@@ -30,15 +31,50 @@ namespace Google.Cloud.Diagnostics.Common
     internal sealed class SimpleManagedTracer : IManagedTracer
     {
         /// <summary>
-        /// A class that represents a running trace span.
+        /// A class that represents a trace span.
         /// </summary>
-        private class Span : IDisposable
+        public sealed class Span : ISpan
         {
+            internal TraceSpan TraceSpan { get; private set; }
+
             private readonly SimpleManagedTracer _tracer;
-            public Span(SimpleManagedTracer tracer) { _tracer = tracer; }
+
+            private readonly object _mutex = new object();
+
+            internal Span(SimpleManagedTracer tracer, TraceSpan traceSpan)
+            {
+                _tracer = GaxPreconditions.CheckNotNull(tracer, nameof(tracer));
+                TraceSpan = GaxPreconditions.CheckNotNull(traceSpan, nameof(traceSpan));
+            }
 
             /// <summary> Ends the current span.</summary>
-            public void Dispose() => _tracer.EndSpan();
+            public void Dispose()
+            {
+                lock (_mutex)
```

Comment:
Hmm, I don't like having to lock here. Would it be possible to have some internal representation of the `EndTime` that we can do an `Interlocked.CompareExchange` operation on to set it only if it is not set and determine if that succeeded.